XMonad/WorkspaceLayout/Core.hs

```diff
@@ -0,0 +1,60 @@
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE NamedFieldPuns             #-}
+{-# LANGUAGE OverloadedLabels           #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TupleSections              #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# OPTIONS_GHC -Wall -Werror #-}
+
+module XMonad.WorkspaceLayout.Core where
+
+import           Prelude                      hiding (span)
+
+import           Control.Category             ((>>>))
+import           Data.Function                (on)
+import           Data.List                    (elemIndex)
+import           GHC.Generics                 (Generic)
+import           XMonad                       hiding (config, modify, state,
+                                               trace, workspaces)
+import           XMonad.Hooks.StatusBar.PP    (PP (..))
+import           XMonad.StackSet              (tag)
+import           XMonad.Util.WorkspaceCompare (mkWsSort)
+
+
+-- |
+--
+-- Encompasses information needed to render a workspace layout
+data WorkspaceLayoutView = WSLView
+  { label        :: String
+  , neighborhood :: [WorkspaceId]
+  , toName       :: WorkspaceId -> String
+  } deriving (Generic)
+
+
+render :: WorkspaceLayoutView -> PP
+render (WSLView { neighborhood, toName, label }) =
+
+  withLabel . withNameTransform . withNeighborhood $ def
+
+  where
+
+  withNameTransform pp = pp
+    { ppCurrent = toName
+    , ppHidden = toName
+    , ppHiddenNoWindows = toName
+    }
```

Comment:
This completely throws away the old values of `pp` which is a confusing behavior. I think it should compose better. The same remark goes for the rest of the `with.*` functions

Reply:
I'll go ahead and inline them. They're all applied only to `def :: PP`, so composition doesn't make a difference, but I agree it doesn't read well as-is

Reply:
Now I understood what you're trying to do (I shouldn't read code when I'm tired). Don't do this, it composes poorly. Instead, use `ppRename`. It was included for this purpose exactly: displaying different names in `PP` than what is actually used internally. `X.L.IndependentScreens` should have an example that is similar to what you're trying to do. Also, the same comment goes to `withNeighborhood`.

Reply:
I've changed the code to use `ppRename`, thanks!

It looks like `IndependentScreens` uses `ppSort` internally same as me. Are you saying that instead of producing a `PP` I should produce an `Endo PP` which handles the composition on the user's behalf?

Reply:
If so, I did change it to `PP -> PP` : p